internal/k8s/helpers.go

```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
 	}
 	return false
 }
+
+// IsObjectEqual checks if objects received during update are equal.
+//
+// Make an attempt to avoid comparing full objects since it can be very CPU intensive.
+// Prefer comparing Generation when only interested in spec changes.
+func IsObjectEqual(old, new interface{}) (bool, error) {
+
+	// Fast path for any object: when ResourceVersions are equal, the objects are equal.
+	// NOTE: This optimizes the case when controller-runtime executes full sync and sends updates for all objects.
+	if isResourceVersionEqual(old, new) {
+		return true, nil
+	}
+
+	switch old := old.(type) {
+
+	// Fast path for objects that implement Generation and where only spec changes matter.
```

Comment:
This seems like a reasonable comparison for this set of resources given that status/annotation/label changes shouldn't need to trigger rebuilds for them.